# On the Client side

Use the 'right' venv for the ipynb kernel (Might need to first `Python: Select Interpreter`, before the venv is proposed as an available kernel).

**Install the packages for the demo in the venv:**

If you already installed an environment, and just wanna check if you have all the required packages:

In [1]:
!pip freeze -r ../requirements.txt

clearml==1.16.4
clearml-agent==1.9.1
scikit-learn==1.5.2
pandas==2.2.2
# from task
joblib==1.4.2
matplotlib==3.9.2
numpy==2.1.1

# for ipynb
ipywidgets==8.1.5
## The following requirements were added by pip freeze:
asttokens==2.4.1
attrs==23.2.0
certifi==2024.8.30
charset-normalizer==3.3.2
comm==0.2.2
contourpy==1.3.0
cycler==0.12.1
debugpy==1.8.5
decorator==5.1.1
distlib==0.3.8
exceptiongroup==1.2.2
executing==2.1.0
fastjsonschema==2.20.0
filelock==3.16.0
fonttools==4.53.1
furl==2.1.3
idna==3.8
ipykernel==6.29.5
ipython==8.27.0
jedi==0.19.1
jsonschema==4.23.0
jsonschema-specifications==2023.12.1
jupyter_client==8.6.2
jupyter_core==5.7.2
jupyterlab_widgets==3.0.13
kiwisolver==1.4.7
matplotlib-inline==0.1.7
nbformat==5.10.4
nest-asyncio==1.6.0
orderedmultidict==1.0.1
packaging==24.1
parso==0.8.4
pathlib2==2.3.7.post1
pexpect==4.9.0
pillow==10.4.0
platformdirs==4.3.2
prompt_toolkit==3.0.47
psutil==5.9.8
ptyprocess==0.7.0
pure_eval==0.2.3
Pygments==2.18.0
PyJWT==2.8.0
pyparsing==3.1.4
pyt

If you notice any:
```bash
WARNING: Requirement file [../requirements.txt] contains pkg_name, but package 'pkg_name' is not installed
```
Then, you are missing packages that will be used in the demo at some points.

**To install the rquired packages:**

Either call:
```bash
make install-env
```
Or directly run:

In [ ]:
# %pip install -r requirements.txt

> <svg style="line-height: 1; height: 1.5em; width: 1.5em; vertical-align: middle;" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--!Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free Copyright 2024 Fonticons, Inc.--><path fill='grey' d="M256 32c14.2 0 27.3 7.5 34.5 19.8l216 368c7.3 12.4 7.3 27.7 .2 40.1S486.3 480 472 480L40 480c-14.3 0-27.6-7.7-34.7-20.1s-7-27.8 .2-40.1l216-368C228.7 39.5 241.8 32 256 32zm0 128c-13.3 0-24 10.7-24 24l0 112c0 13.3 10.7 24 24 24s24-10.7 24-24l0-112c0-13.3-10.7-24-24-24zm32 224a32 32 0 1 0 -64 0 32 32 0 1 0 64 0z"/></svg> WARNING:\
> Using:
> ```python
> !the_cmd
> ```
> Resulted in weird behaviours: agent present to closing down and could not take on any task from the queue
> Instead using:
> ```python
> %%bash
> the_cmd
> ```
> Resulted in the correct behaviours

# Exploring agents

First, let's check if any agent is running:

In [ ]:
%%bash
clearml-agent list

workers: []



With no agent running, you should get the following output:
```bash
workers: []
```


**To launch an agent:**\
*(with name `clearml-agent-bengio`, id `clearml-agent-bengio:gpu0`, on gpu 0 with queue `gpu_dithering`)*

Either call:
```bash
make launch-agent
```
Or directly call:

In [ ]:
%%bash
CLEARML_WORKER_NAME=clearml-agent-bengio CLEARML_WORKER_ID=clearml-agent-bengio:gpu0 clearml-agent daemon --detached --queue gpu_dithering --create-queue --gpus 0

Current configuration (clearml_agent v1.9.1, location: /home/blissfox-server/clearml.conf):
----------------------
agent.worker_id = clearml-agent-bengio:gpu0
agent.worker_name = clearml-agent-bengio
agent.force_git_ssh_protocol = false
agent.python_binary = /usr/bin/python3.10
agent.package_manager.type = pip
agent.package_manager.pip_version.0 = <20.2 ; python_version < '3.10'
agent.package_manager.pip_version.1 = <22.3 ; python_version >\= '3.10' and python_version <\= '3.11'
agent.package_manager.pip_version.2 = >\=23,<24.3 ; python_version >\= '3.12'
agent.package_manager.system_site_packages = false
agent.package_manager.force_upgrade = false
agent.package_manager.conda_channels.0 = pytorch
agent.package_manager.conda_channels.1 = conda-forge
agent.package_manager.conda_channels.2 = nvidia
agent.package_manager.conda_channels.3 = defaults
agent.package_manager.priority_optional_packages.0 = pygobject
agent.package_manager.torch_nightly = false
agent.package_manager.poetry_files_f

<details>
<summary>Output</summary>

```bash
Current configuration (clearml_agent v1.9.1, location: /home/blissfox-server/clearml.conf):
----------------------
agent.worker_id = clearml-agent-bengio:gpu0
agent.worker_name = clearml-agent-bengio
agent.force_git_ssh_protocol = false
agent.python_binary = /usr/bin/python3.10
agent.package_manager.type = pip
agent.package_manager.pip_version.0 = <20.2 ; python_version < '3.10'
agent.package_manager.pip_version.1 = <22.3 ; python_version >\= '3.10' and python_version <\= '3.11'
agent.package_manager.pip_version.2 = >\=23,<24.3 ; python_version >\= '3.12'
agent.package_manager.system_site_packages = false
agent.package_manager.force_upgrade = false
agent.package_manager.conda_channels.0 = pytorch
agent.package_manager.conda_channels.1 = conda-forge
agent.package_manager.conda_channels.2 = nvidia
agent.package_manager.conda_channels.3 = defaults
agent.package_manager.priority_optional_packages.0 = pygobject
agent.package_manager.torch_nightly = false
agent.package_manager.poetry_files_from_repo_working_dir = false
agent.venvs_dir = /home/blissfox-server/.clearml/venvs-builds
agent.venvs_cache.max_entries = 10
agent.venvs_cache.free_space_threshold_gb = 2.0
agent.venvs_cache.path = ~/.clearml/venvs-cache
agent.vcs_cache.enabled = true
agent.vcs_cache.path = /home/blissfox-server/.clearml/vcs-cache
...
| 6b55ab8e3be4482eb53acfed3e4037a5 | gpu_dithering |       |
+----------------------------------+---------------+-------+

Running CLEARML-AGENT daemon in background mode, writing stdout/stderr to /tmp/.clearml_agent_daemon_out47ajuksj.txt
```

</details>

<details>
<summary>About the cmd & arguments</summary>

From [ClearML Agent CLI](https://clear.ml/docs/latest/docs/clearml_agent/clearml_agent_ref/):
|||
|:---|:---|
|`daemon`|command to spin up an agent on any machine|
|`--create-queue`|If the queue name provided with `--queue` does not exist in the server, create it on-the-fly and use it.|
|`--detached`|Run agent in the background. The `clearml-agent` command returns immediately.|
|`--gpus`|If running in Docker mode (see the `--docker` option), specify the active GPUs for the Docker containers to use. These are the same GPUs set in the `NVIDIA_VISIBLE_DEVICES` environment variable. For example:<ul><li>`--gpus 0`</li><li>`--gpu 0,1,2`</li><li>`--gpus all`</li></ul>|
|`--queue`|Specify the queues which the worker is listening to. The values can be any combination of:<ul><li>One or more queue IDs</li><li>One or more queue names</li><li>`default` indicating the default queue</li></ul>
</details>

<details>
<summary>On the GUI:</summary>

* Before:\
    ![](./../img/Workers_null.png)
* After:\
    ![](./../img/Workers_bengio0_0.png)\
    ![](./../img/Workers_bengio0_1.png)

</details>

Let's check out again the agent available from the CLI:

In [ ]:
%%bash
clearml-agent list

workers:
- company:
    id: d1bd92a3b039400cbafc60a7a5b1e52b
    name: clearml
  id: clearml-agent-bengio:gpu0
  ip: 172.25.1.1
  key: worker_d1bd92a3b039400cbafc60a7a5b1e52b_0f358a174884f2c34ef7def624d3e1a0_clearml-agent-bengio:gpu0
  last_activity_time: '2024-09-28T10:27:23.587122+00:00'
  last_report_time: '2024-09-28T10:27:23.587122+00:00'
  queues:
  - id: 6b55ab8e3be4482eb53acfed3e4037a5
    name: gpu_dithering
    num_tasks: 0
  register_time: '2024-09-28T10:27:20.249917+00:00'
  register_timeout: 600
  system_tags: []
  tags: []
  user:
    id: 0f358a174884f2c34ef7def624d3e1a0
    name: RR5555



<details>
<summary>Output</summary>

```bash
workers:
- company:
    id: d1bd92a3b039400cbafc60a7a5b1e52b
    name: clearml
  id: clearml-agent-bengio:gpu0
  ip: 172.25.1.1
  key: worker_d1bd92a3b039400cbafc60a7a5b1e52b_0f358a174884f2c34ef7def624d3e1a0_clearml-agent-bengio:gpu0
  last_activity_time: '2024-09-16T11:46:44.589928+00:00'
  queues:
  - id: 6b55ab8e3be4482eb53acfed3e4037a5
    name: gpu_dithering
    num_tasks: 0
  register_time: '2024-09-16T11:46:44.589928+00:00'
  register_timeout: 600
  system_tags: []
  tags: []
  user:
    id: 0f358a174884f2c34ef7def624d3e1a0
    name: RR5555
```
</details>

In [ ]:
%%bash
docker ps

In [21]:
%%bash
ls -a /tmp/ | grep -i 'clearml\|requirements'
# OR equivalently:
# ls -a /tmp/ | grep -E -i 'clearml|requirements'

.clearmlagent_83894_2cktpmnu.tmp
.clearml_agent.bkmgjexz.cfg
.clearml_agent_daemon_out0v1et8_d.txt
requirements_uu2a4f6i.txt


```bash
.clearmlagent_83894_2cktpmnu.tmp
.clearml_agent.bkmgjexz.cfg
.clearml_agent_daemon_out0v1et8_d.txt
requirements_uu2a4f6i.txt
```

**Stop the agent:**

In [ ]:
%%bash
clearml-agent daemon --stop clearml-agent-bengio:gpu0

In [18]:
%%bash
clearml-agent list

workers:
- company:
    id: d1bd92a3b039400cbafc60a7a5b1e52b
    name: clearml
  id: clearml-agent-bengio:gpu0
  ip: 172.25.1.1
  key: worker_d1bd92a3b039400cbafc60a7a5b1e52b_0f358a174884f2c34ef7def624d3e1a0_clearml-agent-bengio:gpu0
  last_activity_time: '2024-09-29T11:03:33.568803+00:00'
  last_report_time: '2024-09-29T11:03:33.568803+00:00'
  queues:
  - id: 6b55ab8e3be4482eb53acfed3e4037a5
    name: gpu_dithering
    num_tasks: 0
  register_time: '2024-09-29T10:25:05.394822+00:00'
  register_timeout: 600
  system_tags: []
  tags: []
  user:
    id: 0f358a174884f2c34ef7def624d3e1a0
    name: RR5555



## Launch an exp. from a public repo.

We will:
* Launch an agent/worker on a brand new queue: it will wait for a load (created a container? Or container only under load?)
* Launch a task from the CLI calling a script from a public GitHub repo
* Stop the agent

In [ ]:
%%bash
CLEARML_WORKER_NAME=clearml-agent-bengio CLEARML_WORKER_ID=clearml-agent-bengio:gpu0 clearml-agent daemon --detached --queue gpu_dithering --create-queue --gpus 0
clearml-task --project Demo --name clearml_test --script test.py --repo https://github.com/RR5555/clearml_demo.git --branch main --queue gpu_dithering --docker rr5555/dith:cuda12.0-agent
clearml-agent daemon --stop clearml-agent-bengio:gpu0

## Check out on Tasks from Python



Unfortunately, it seems that, for now (2024/09/18), there is no CLI cmd to monitor the tasks.\
No such things as `clearml-task list`.

We thus have to fall back to the python interface: [ClearML -- Accessing Tasks](https://clear.ml/docs/latest/docs/clearml_sdk/task_sdk/#accessing-tasks)

[ClearML -- Task get_status](https://clear.ml/docs/latest/docs/references/sdk/task/#get_status):
* TaskStatusEnum: [“created”, “in_progress”, “stopped”, “closed”, “failed”, “completed”, “queued”, “published”, “publishing”, “unknown”]

In [3]:
from clearml import Task

> Do not worry about reruning the import cell above as, if the python module is already loaded, it won't be loaded twice.

Get all the tasks `in_progress` in the project `Demo`:

In [4]:
in_progress_task_list = Task.get_tasks(
    task_ids=None,  # type Optional[Sequence[str]]
    project_name='Demo',  # Optional[str]
    task_name=None,  # Optional[str]
    allow_archived=True, # [bool]
    task_filter={
        'status': ('in_progress',)
    },  # Optional[Dict]#
    # tasks with tag `included_tag` or without tag `excluded_tag`
    tags=None #['included_tag', '-excluded_tag'] 
)
print(f"Number of `in_progress` task: {len(in_progress_task_list)}")
print(in_progress_task_list)

[]


Get the tasks `completed` in project `Demo` ordered by `last_update`:

In [6]:
completed_task_list = Task.get_tasks(
    task_ids=None,  # type Optional[Sequence[str]]
    project_name='Demo',  # Optional[str]
    task_name=None,  # Optional[str]
    allow_archived=True, # [bool]
    task_filter={
        'status': ('completed',),
        'order_by': ('last_update',)
    },  # Optional[Dict]#
    # tasks with tag `included_tag` or without tag `excluded_tag`
    tags=None #['included_tag', '-excluded_tag'] 
)
print(f"Number of `completed` task: {len(completed_task_list)}")
print(completed_task_list)

Number of `completed` task: 9
[<clearml.task.Task object at 0x7cb998089ea0>, <clearml.task.Task object at 0x7cb9c5555900>, <clearml.task.Task object at 0x7cb98d5bb130>, <clearml.task.Task object at 0x7cb98d5bb3a0>, <clearml.task.Task object at 0x7cb98d5e4820>, <clearml.task.Task object at 0x7cb98d5e50c0>, <clearml.task.Task object at 0x7cb98d5e59c0>, <clearml.task.Task object at 0x7cb98d5e6140>, <clearml.task.Task object at 0x7cb98d5e6890>]


### Explore a task INFO panel information

Access the information from the `INFO` panel of the clearml GUI from the last task in the previous list:

In [7]:
_task = completed_task_list[-1]

If you want to explore the fields of the task, just list them `obj.__dir__()` or `dir(obj)`:
```python
print(_task.__dir__())
print(tuple(_field for _field in _task.__dir__() if _field.find('get')>-1))
```

In [ ]:
# print(_task.__dir__())
# print(tuple(_field for _field in _task.__dir__() if _field.find('get')>-1))
print(tuple(_field for _field in _task.__dir__() if _field.find('name')>-1))

Let's look at some of the task informative properties:

In [8]:
from IPython.display import display, Markdown
from collections import OrderedDict

_cmds = OrderedDict((
            ('_task.get_reported_console_output()', _task.get_reported_console_output()),
            ('_task._get_runtime_properties()', _task._get_runtime_properties()),
            ('_task.get_tags()', _task.get_tags()),
            ('_task.get_script()', _task.get_script()),
            ('_task.comment', _task.comment),
            ('_task.status', _task.status),
            ('_task.get_parameters()', _task.get_parameters()),
            ('_task.get_task_output_log_web_page(_task.id)', _task.get_task_output_log_web_page(_task.id)),
            ('_task.name', _task.name),
            ('_task.id', _task.id),
            ('_task._project_name', _task._project_name),
            ('_task.get_output_destination()', _task.get_output_destination())
))

_md_table="|Cmd|Result|\n|:---|:---|\n"
for _k,_v in _cmds.items():
    _md_table += (f"|`{_k}`|`{_v}`|\n")

# print(_md_table)

# display_markdown(_md_table)
display(Markdown(_md_table))


|Cmd|Result|
|:---|:---|
|`_task.get_reported_console_output()`|`['Process completed successfully']`|
|`_task._get_runtime_properties()`|`{'_CLEARML_TASK': True, '_exec_agent_hostname': 'bengio', 'CLEARML VERSION': 'clearml-1.16.4', 'CLI': 'test.py', 'progress': '0', 'platform': 'linux', 'python_version': '3.10.12', 'python_exec': '/home/blissfox-server/.clearml/venvs-builds/3.10/bin/python', 'OS': 'Linux-6.8.0-40-generic-x86_64-with-glibc2.35', 'processor': 'x86_64', 'cpu_cores': 56, 'memory_gb': 62.8, 'hostname': 'bengio', 'gpu_count': 1, 'gpu_type': 'NVIDIA TITAN X (Pascal)', 'gpu_memory': '12GB', 'gpu_driver_version': '535.183.01', 'gpu_driver_cuda_version': '12.2'}`|
|`_task.get_tags()`|`[]`|
|`_task.get_script()`|`{'working_dir': '.', 'entry_point': 'test.py', 'branch': 'main', 'repository': 'https://github.com/RR5555/clearml_demo.git'}`|
|`_task.comment`|`Auto-generated at 2024-09-17 12:23:52 UTC by blissfox-server@bengio`|
|`_task.status`|`completed`|
|`_task.get_parameters()`|`{}`|
|`_task.get_task_output_log_web_page(_task.id)`|`https://app.clearml.blissfox.xyz/projects/*/experiments/5812d0aa235f4858aa1e6c77925f0d3c/output/log`|
|`_task.name`|`clearml_test`|
|`_task.id`|`5812d0aa235f4858aa1e6c77925f0d3c`|
|`_task._project_name`|`('ce6d125a264149bc89e8e62fb1dd0d7a', None)`|
|`_task.get_output_destination()`|`None`|


Output example:

<div markdown=1 style=" overflow-x: auto;">

|Cmd|Result|
|:---|:---|
|`_task.get_reported_console_output()`|`['Process completed successfully']`|
|`_task._get_runtime_properties()`|`{'_CLEARML_TASK': True, '_exec_agent_hostname': 'bengio', 'CLEARML VERSION': 'clearml-1.16.4', 'CLI': 'test.py', 'progress': '0', 'platform': 'linux', 'python_version': '3.10.12', 'python_exec': '/home/blissfox-server/.clearml/venvs-builds/3.10/bin/python', 'OS': 'Linux-6.8.0-40-generic-x86_64-with-glibc2.35', 'processor': 'x86_64', 'cpu_cores': 56, 'memory_gb': 62.8, 'hostname': 'bengio', 'gpu_count': 1, 'gpu_type': 'NVIDIA TITAN X (Pascal)', 'gpu_memory': '12GB', 'gpu_driver_version': '535.183.01', 'gpu_driver_cuda_version': '12.2'}`|
|`_task.get_tags()`|`[]`|
|`_task.get_script()`|`{'working_dir': '.', 'entry_point': 'test.py', 'branch': 'main', 'repository': 'https://github.com/RR5555/clearml_demo.git'}`|
|`_task.comment`|`Auto-generated at 2024-09-17 12:23:52 UTC by blissfox-server@bengio`|
|`_task.status`|`completed`|
|`_task.get_parameters()`|`{}`|
|`_task.get_task_output_log_web_page(_task.id)`|`https://app.clearml.blissfox.xyz/projects/*/experiments/5812d0aa235f4858aa1e6c77925f0d3c/output/log`|
|`_task.name`|`clearml_test`|
|`_task.id`|`5812d0aa235f4858aa1e6c77925f0d3c`|
|`_task._project_name`|`('ce6d125a264149bc89e8e62fb1dd0d7a', None)`|
|`_task.get_output_destination()`|`None`|
</div>

In [11]:
print(_task.data)

<tasks.Task: {
    "id": "5812d0aa235f4858aa1e6c77925f0d3c",
    "name": "clearml_test",
    "user": "0f358a174884f2c34ef7def624d3e1a0",
    "company": "d1bd92a3b039400cbafc60a7a5b1e52b",
    "type": "training",
    "status": "completed",
    "comment": "Auto-generated at 2024-09-17 12:23:52 UTC by blissfox-server@bengio",
    "created": "2024-09-17 12:23:52.352000+00:00",
    "started": "2024-09-17 12:23:54.108000+00:00",
    "completed": "2024-09-17 12:28:29.926000+00:00",
    "active_duration": 275,
    "project": "ce6d125a264149bc89e8e62fb1dd0d7a",
    "output": {},
    "execution": {
        "queue": "6b55ab8e3be4482eb53acfed3e4037a5",
        "parameters": {},
        "model_desc": {},
        "model_labels": {},
        "artifacts": []
    },
    "container": {
        "image": "rr5555/dith:cuda12.0-agent",
        "arguments": "",
        "setup_shell_script": ""
    },
    "models": {
        "input": [],
        "output": []
    },
    "script": {
        "binary": "python",


In [ ]:
print(dir(_task.data))

In [65]:
# https://stackoverflow.com/a/50215584/8612123

def print_table(data, cols, wide):
    '''Prints formatted data on columns of given width.'''
    n, r = divmod(len(data), cols)
    pat = '{{:{}}}'.format(wide)
    line = '\n'.join(pat * cols for _ in range(n))
    last_line = pat * r
    print(line.format(*data))
    print(last_line.format(*data[n*cols:]))

In [66]:
print_table(dir(_task.data), 4, len(dir(_task.data))//4+1)

__class__                 __delattr__               __dict__                  __dir__                   
__doc__                   __eq__                    __format__                __ge__                    
__getattribute__          __gt__                    __hash__                  __init__                  
__init_subclass__         __le__                    __lt__                    __module__                
__ne__                    __new__                   __reduce__                __reduce_ex__             
__repr__                  __setattr__               __sizeof__                __str__                   
__subclasshook__          __weakref__               _data_props_list          _get_data_props           
_property_active_duration _property_comment         _property_company         _property_completed       
_property_configuration   _property_container       _property_created         _property_execution       
_property_hyperparams     _property_id              _pr

In [67]:
def get_instance_field(_obj, _class):
	return tuple(_field for _field in dir(_obj) if not callable(getattr(_obj, _field)) and isinstance(getattr(_obj, _field), _class))

In [70]:
from pprint import pprint
import datetime

In [71]:
pprint(get_instance_field(_task.data, datetime.datetime))

('_property_completed',
 '_property_created',
 '_property_last_change',
 '_property_last_update',
 '_property_last_worker_report',
 '_property_started',
 '_property_status_changed',
 'completed',
 'created',
 'last_change',
 'last_update',
 'last_worker_report',
 'started',
 'status_changed')


In [72]:
pprint(get_instance_field(_task.data, str))

('__doc__',
 '__module__',
 '_property_comment',
 '_property_company',
 '_property_id',
 '_property_last_worker',
 '_property_name',
 '_property_project',
 '_property_status_message',
 '_property_status_reason',
 '_property_user',
 'comment',
 'company',
 'id',
 'last_worker',
 'name',
 'project',
 'status_message',
 'status_reason',
 'user')


In [73]:
pprint(get_instance_field(_task.data, dict))

('__dict__',
 '_data_props_list',
 '_property_configuration',
 '_property_container',
 '_property_hyperparams',
 '_property_last_metrics',
 '_property_runtime',
 '_schema',
 'configuration',
 'container',
 'hyperparams',
 'last_metrics',
 'runtime')


In [75]:
nb_fields = len(tuple(_field for _field in dir(_task.data) if not callable(getattr(_task.data, _field))))

print(nb_fields, len(_task.data.__dir__()))

72 101


#### Play with time

In [21]:
import datetime

In [68]:
datetime_fields = get_instance_field(_task.data, datetime.datetime)

print_table(datetime_fields, 1, len(datetime_fields)//1+1)

_property_completed
_property_created
_property_last_change
_property_last_update
_property_last_worker_report
_property_started
_property_status_changed
completed      
created        
last_change    
last_update    
last_worker_report
started        
status_changed 



In [22]:
print(f"`type(_task.data.created)`: {type(_task.data.created)}")
print(f"Created at: {_task.data.created}")

`type(_task.data.created)`: <class 'datetime.datetime'>
Created at: 2024-09-17 12:23:52.352000+00:00


To compare dates in `datetime.datetime`, we can use:
* Now: `datetime.datetime.now(tz=_task.data.created.tzinfo)`
* Deltas (7 days): `datetime.timedelta(days=7)`
* 7 days ago: `datetime.datetime.now(tz=_task.data.created.tzinfo) - datetime.timedelta(days=7)`

In [24]:
print(f"Now: {datetime.datetime.now(tz=_task.data.created.tzinfo)}")
print(f"7 days ago: {datetime.datetime.now(tz=_task.data.created.tzinfo) - datetime.timedelta(days=7)}")

Now: 2024-09-26 13:57:00.594638+00:00
7 days ago: 2024-09-19 13:57:00.594898+00:00


In [27]:
print(f"Was the last exp. created in the last 7 days: {'Yes' if (_task.data.created> datetime.datetime.now(tz=_task.data.created.tzinfo) - datetime.timedelta(days=7)) else 'No'}")

Was the last exp. created in the last 7 days: No


In [37]:
def get_exps_from_last(_list, _days):
	return tuple(_task for _task in _list if _task.data.created> datetime.datetime.now(tz=_task.data.created.tzinfo) - datetime.timedelta(days=_days))

In [38]:
print(f"Exp. in the last two months: {len(get_exps_from_last(completed_task_list, 60))}")
print(f"Exp. in the last year: {len(get_exps_from_last(completed_task_list, 365))}")

Exp. in the last two months: 1
Exp. in the last year: 9


## Launch an exp. from a private repo.

We will:
* Launch an agent/worker on a brand new queue: it will wait for a load (created a container? Or container only under load?)
* Launch a task from the CLI calling a script from a public GitHub repo
* Stop the agent

**Create a Personal Access Token (PAT)**

<details>
<summary>Click to expand</summary>

To create a PAT, heads to `Settings`>`Developer settings`>`Personnal access tokens`>`Fine-grained tokens`, or click on the link below:\
https://github.com/settings/personal-access-tokens/new

This will lead you to this page:\
<img src="../img/new_PAT0.png" style="">\
Enter the `Token name`, `Expiration` & `Description`.

Select the repos you want the token to apply for:\
<img src="../img/new_PAT1.png" style="">

Choose the `Permissions`:
<img src="../img/new_PAT2.png" style="">

I only added the `Contents`:`Access: Read-only` (in addition to the default compulsory Metadata access):\
<img src="../img/new_PAT3.png" style="">

You can check out the permissions you will give in `Overview`:\
<img src="../img/new_PAT4.png" style="">\
Then confirm by `Generate token`.

You can now copy the token to where you will need it (you can only do it now, if you did not, you will have to regenerate the token, to copy the new one):\
<img src="../img/new_PAT5.png" style="">

</details>

To pass down the PAT, either configure in the default clearml-agent configs, or pass it down at agent creation:

In [ ]:
%%bash
CLEARML_AGENT_GIT_USER="rr5555" CLEARML_AGENT_GIT_PASS=`cat ../pat.txt` CLEARML_WORKER_NAME=clearml-agent-bengio CLEARML_WORKER_ID=clearml-agent-bengio:gpu1 clearml-agent daemon --detached --queue gpu_dithering --create-queue --gpus 1
clearml-task --project Demo --name clearml_test --script test.py --repo https://github.com/RR5555/clearml_demo_private.git --branch main --queue gpu_dithering --docker rr5555/dith:cuda12.0-agent
clearml-agent daemon --stop clearml-agent-bengio:gpu1

ClearML launch - launch any codebase on remote machine running clearml-agent
Creating new task
Repository Detected
{
  "repository": "https://github.com/RR5555/clearml_demo_private.git",
  "branch": "main",
  "working_dir": ".",
  "entry_point": "test.py"
}
Base docker image: {'image': 'rr5555/dith:cuda12.0-agent', 'args': None, 'bash_script': None}
New task created id=d12f6da8220c46269c939c2040e71f97
Task id=d12f6da8220c46269c939c2040e71f97 sent for execution on queue gpu_dithering
Execution log at: https://app.clearml.blissfox.xyz/projects/ce6d125a264149bc89e8e62fb1dd0d7a/experiments/d12f6da8220c46269c939c2040e71f97/output/log


Trying to directly run the experiment will fail as the repo is private and no PAT for it is passed:

In [ ]:
%%bash
CLEARML_WORKER_NAME=clearml-agent-bengio CLEARML_WORKER_ID=clearml-agent-bengio:gpu0 clearml-agent daemon --detached --queue gpu_dithering --create-queue --gpus 0
clearml-task --project Demo --name clearml_test --script test.py --repo https://github.com/RR5555/clearml_demo_private.git --branch main --queue gpu_dithering --docker rr5555/dith:cuda12.0-agent
clearml-agent daemon --stop clearml-agent-bengio:gpu0

It should also be noted that passing it down to `clearml-task` either as env. vars or docker env. vars will fail too:

In [16]:
%%bash
CLEARML_WORKER_NAME=clearml-agent-bengio CLEARML_WORKER_ID=clearml-agent-bengio:gpu0 clearml-agent daemon --detached --queue gpu_dithering --create-queue --gpus 0
CLEARML_AGENT_GIT_USER="rr5555" CLEARML_AGENT_GIT_PASS=`cat ../pat.txt` clearml-task --project Demo --name clearml_test --script test.py --repo https://github.com/RR5555/clearml_demo_private.git --branch main --queue gpu_dithering --docker rr5555/dith:cuda12.0-agent
clearml-agent daemon --stop clearml-agent-bengio:gpu0

ClearML launch - launch any codebase on remote machine running clearml-agent
Creating new task
Repository Detected
{
  "repository": "https://github.com/RR5555/clearml_demo_private.git",
  "branch": "main",
  "working_dir": ".",
  "entry_point": "test.py"
}
Base docker image: {'image': 'rr5555/dith:cuda12.0-agent', 'args': None, 'bash_script': None}
New task created id=aacc308dc02941c7aa9fd33ed0eee746
Task id=aacc308dc02941c7aa9fd33ed0eee746 sent for execution on queue gpu_dithering
Execution log at: https://app.clearml.blissfox.xyz/projects/ce6d125a264149bc89e8e62fb1dd0d7a/experiments/aacc308dc02941c7aa9fd33ed0eee746/output/log


In [ ]:
%%bash
CLEARML_WORKER_NAME=clearml-agent-bengio CLEARML_WORKER_ID=clearml-agent-bengio:gpu0 clearml-agent daemon --detached --queue gpu_dithering --create-queue --gpus 0
clearml-task --project Demo --name clearml_test --script test.py --repo https://github.com/RR5555/clearml_demo_private.git --branch main --queue gpu_dithering --docker rr5555/dith:cuda12.0-agent --docker_args "-v CLEARML_AGENT_GIT_USER:rr5555 -v CLEARML_AGENT_GIT_PASS:`cat ../pat.txt`"
clearml-agent daemon --stop clearml-agent-bengio:gpu0

```bash
make launch-agents
make pipeline-from-functions
make pipeline-from-decorator

make create-tasks-for-pipeline
make pipeline-from-tasks

make stop-agents
```

https://clear.ml/docs/latest/docs/configs/clearml_conf/

agent.docker_use_activated_venv (bool)